lib/collection/src/collection_manager/holders/segment_holder.rs

```diff
@@ -527,6 +510,85 @@ impl<'s> SegmentHolder {
         }
         Ok(removed_points)
     }
+
+    fn find_duplicated_points(&self) -> OperationResult<HashMap<SegmentId, Vec<PointIdType>>> {
+        let segments = self
+            .segments
+            .iter()
+            .map(|(&segment_id, locked_segment)| (segment_id, locked_segment.get()))
+            .collect_vec();
+        let locked_segments = BTreeMap::from_iter(
+            segments
+                .iter()
+                .map(|(segment_id, locked_segment)| (*segment_id, locked_segment.read())),
+        );
+        let mut iterators = BTreeMap::from_iter(
+            locked_segments
+                .iter()
+                .map(|(segment_id, locked_segment)| (*segment_id, locked_segment.iter_points())),
+        );
+
+        // heap contains the current iterable point id from each segment
+        let mut heap = iterators
+            .iter_mut()
+            .filter_map(|(&segment_id, iter)| {
+                iter.next().map(|point_id| DedupPoint {
+                    segment_id,
+                    point_id,
+                })
+            })
+            .collect::<BinaryHeap<_>>();
+
+        let mut last_point_id_opt = None;
+        let mut last_segment_id_opt = None;
+        let mut last_point_version_opt = None;
+        let mut points_to_remove: HashMap<SegmentId, Vec<PointIdType>> = Default::default();
+
+        while let Some(mut entry) = heap.pop() {
+            let point_id = entry.point_id;
+            let segment_id = entry.segment_id;
+            if let Some(next_point_id) = iterators.get_mut(&segment_id).and_then(|i| i.next()) {
+                entry.point_id = next_point_id;
+                heap.push(entry);
+            }
+
+            if last_point_id_opt == Some(point_id) {
+                let last_point_id = last_point_id_opt.unwrap();
+                let last_segment_id = last_segment_id_opt.unwrap();
+
+                let point_version = locked_segments[&segment_id].point_version(point_id);
+                let last_point_version = if let Some(last_point_version) = last_point_version_opt {
+                    last_point_version
+                } else {
+                    let version = locked_segments[&last_segment_id].point_version(last_point_id);
+                    last_point_version_opt = Some(version);
+                    version
+                };
+
+                // choose newer version between point_id and last_point_id
+                if last_point_version > point_version {
```

Comment:
🪡: I'd swap operands in the point version comparison.

```suggestion
                if point_version < last_point_version {
```

(Or `last_point_version <= point_version` + swap branches.)

IMO, it's easier to read conditionals when they are structured like "if _CURRENT_ point is... then remove _CURRENT_ point", instead of "if _LAST_ point is... then remove _CURRENT_ point"

But this is a Super Nitpick™️.

Reply:
Np